### PR TITLE
Network Manager Invite and Claim

### PR DESCRIPTION
**Pull request by hemant10yadav**
https://dimagi.atlassian.net/browse/CCCT-456

Screenshot 2024-08-14 at 12 54 54
Screenshot 2024-08-14 at 13 00 29
Screenshot 2024-08-14 at 13 00 07
Screenshot 2024-08-14 at 12 55 14

### CODE REVIEW

File: commcare_connect/templates/opportunity/opportunity_list.html
Comment (calellowitz) [reply commcare_connect/templates/opportunity/opportunity_list.html]: strong plus one to this

File: commcare_connect/opportunity/views.py
Comment (calellowitz) [reply commcare_connect/opportunity/views.py:132]: agree with this. I made a comment on the other one, thinking we had decided to get rid of the second model entirely, but agree that generally claimed opportunities should have the organization of the network manager.

File: commcare_connect/program/tests/test_views.py
Comment (calellowitz) [inline commcare_connect/program/tests/test_views.py:20]: It would be good to add tests for the forms, I know the managed opportunity one handles a lot, and especially any validation conditions so both good and bad data. It is great that this includes a lot of happy path tests, but tests that confirm expected failures can also be very useful.

**Review by calellowitz — COMMENTED** [review]
generally looks good. left one comment but otherwise excited to see QA for all this

File: commcare_connect/program/tests/test_views.py
Comment (hemant10yadav) [reply commcare_connect/program/tests/test_views.py:20]: Added

File: commcare_connect/opportunity/views.py
Comment (hemant10yadav) [reply commcare_connect/opportunity/views.py]: Added [fc4ec11](https://github.com/dimagi/commcare-connect/pull/379/commits/fc4ec119c07c654a333af2f0cb15b8c65d07365b)

File: commcare_connect/templates/opportunity/opportunity_list.html
Comment (hemant10yadav) [reply commcare_connect/templates/opportunity/opportunity_list.html]: I have implemented this in all the other views, but this one is a bit tricky because there is already a table (Opportunity list) that includes sorting. If I create another table with Django-tables2, the pagination of the first table will affect the second one. I think I might need to change the entire view to implement Django-tables2. Should I do that?

File: commcare_connect/organization/decorators.py
Comment (sravfeyn) [inline commcare_connect/organization/decorators.py:21]: Shouldn't this be`_request_user_is_program_manager`?

Comment (hemant10yadav) [reply commcare_connect/organization/decorators.py]: Done [84a04d7](https://github.com/dimagi/commcare-connect/pull/379/commits/84a04d76f25496165c9c7b6a5085bc87ccc5284a)

File: commcare_connect/opportunity/views.py
Comment (hemant10yadav) [inline commcare_connect/opportunity/views.py]: This is done to ensure that the same creation flow can be used for both opportunities and managed opportunities. Since a managed opportunity will not have an organization initially, we check the program's organization in the case of a managed opportunity to avoid 404 error.

we don't need to return `opp.managedopportunity` so changed the code little bit here [276fef0](https://github.com/dimagi/commcare-connect/pull/379/commits/276fef074507256f35a10eea4ce4f948cf5be67b)

File: commcare_connect/templates/opportunity/opportunity_list.html
Comment (pxwxnvermx) [reply commcare_connect/templates/opportunity/opportunity_list.html]: I think we are fine with leaving the Opportunity List table for now.

File: commcare_connect/program/models.py
Comment (calellowitz) [inline commcare_connect/program/models.py:18]: I don't think we want this to cascade, because we might still want it to show up for the network managers. I think for now lets `PROTECT` and we can always change it later

File: commcare_connect/templates/program/program_list.html
Comment (calellowitz) [inline commcare_connect/templates/program/program_list.html:22]: nice! this is so much better

File: commcare_connect/opportunity/forms.py
Comment (calellowitz) [inline commcare_connect/opportunity/forms.py:241]: This is no longer true, right? and we can remove the migration that makes org nullable?

File: commcare_connect/opportunity/views.py
Comment (calellowitz) [inline commcare_connect/opportunity/views.py:343]: why does `request.org` no longer work? For all the views in this commit, it would be nice to add the `org_slug` arg to the function signature rather than pulling it from `kwargs`, so `def export_user_visits(request, org_slug, **kwargs):` or similar

**Review by calellowitz — COMMENTED** [review]
left a few questions/comments but looks good

File: commcare_connect/opportunity/views.py
Comment (hemant10yadav) [reply commcare_connect/opportunity/views.py:343]: That was my mistake for not realizing earlier that req.org.org_slug is the same as org_slug. I apologize for that. I will make the necessary changes.

File: commcare_connect/opportunity/forms.py
Comment (hemant10yadav) [reply commcare_connect/opportunity/forms.py:241]: Yeah this is changed in the later PR's

File: commcare_connect/program/models.py
Comment (hemant10yadav) [reply commcare_connect/program/models.py:18]: Changed here ([0924f21](https://github.com/dimagi/commcare-connect/pull/385/commits/0924f21617da12af23fd9b877b3f13bc67d70a42))

As in program invite flow this script again got changed so to avoid conflict did this here

File: commcare_connect/opportunity/views.py
Comment (calellowitz) [reply commcare_connect/opportunity/views.py]: I won't block on this, because I see that we use this pattern elsewhere, but what I was trying to say was that the function based views would be a lot cleaner if they included the arg names explicitly in the signature, rather than using `kwargs` to get the org slug. Like these views for example https://github.com/dimagi/commcare-connect/blob/main/commcare_connect/opportunity/views.py#L348

There should also be a way to do that with the class based views but I don't have an easy example at hand

**Review by calellowitz — APPROVED** [review]
I dont think my last piece of feedback was really followed, but approving to not block release. It would be great if you could create a followup PR after these are merged with that change

File: commcare_connect/opportunity/views.py
Comment (hemant10yadav) [reply commcare_connect/opportunity/views.py]: Ok I will create a new PR for these improvements soon.

**Pull request merged**
